enclave-runtime/src/top_pool_execution.rs

```diff
@@ -162,23 +164,29 @@ where
 			stf_executor.clone(),
 		));
 
-	let block_composer =
-		Arc::new(BlockComposer::new(authority.clone(), state_key, rpc_author, stf_executor));
+	let block_composer = Arc::new(BlockComposer::new(authority.clone(), state_key, rpc_author));
 
-	match yield_next_slot(duration_now(), SLOT_DURATION, latest_onchain_header, &mut LastSlotSeal)?
-	{
+	match yield_next_slot(
+		duration_now(),
+		SLOT_DURATION,
+		latest_parentchain_header,
+		&mut LastSlotSeal,
+	)? {
 		Some(slot) => {
 			let shards = state_handler.list_shards()?;
-			let env = ProposerFactory::new(top_pool_executor, block_composer);
-
-			exec_aura_on_slot::<_, _, SignedSidechainBlock, _, _, _, _>(
-				slot,
-				authority,
+			let env = ProposerFactory::new(top_pool_executor, stf_executor, block_composer);
+
+			let (blocks, opaque_calls) = exec_aura_on_slot::<_, _, SignedSidechainBlock, _, _>(
+				slot, authority, OcallApi, env, shards,
+			)?;
+			// Drop lock as soon as we don't need it anymore.
+			drop(_enclave_write_lock);
```

Comment:
Is the sidechain state altered in `exec_aura_on_slot` only? If so, we should probably acquire the lock inside this `Some(slot) => ` arm, to minimize the time we're holding the lock (or even move it into aura).

Reply:
It depends - do we want to acquire the lock only to write, or do we also want to assure that while processing calls, no one alters the state in the enclave? For example update the latest finalized parentchain header. (though I'm not sure if the enclave lock prevents this).

Reply:
the parentchain block/header syncing and locking is done separately. I think we leave it as is for now - but we might want to come back to this question at some point.